### ISSUE TITLE
DNS返回IPV4数量导致的问题

### ISSUE THREAD

**Issue by qwerttvv** [shadowsocks/shadowsocks-rust#1425]
问题发现过程见这里
https://github.com/pymumu/smartdns/issues/1657

最后猜测是不是代理工具传送dns解析的ip的时候，如果ip太多，比如ipv4达到33个，会导致错误

服务端会显示

ERROR tcp tunnel 127.0.0.1:14050 -> s3-us-west-1.amazonaws.com:443 connect failed, error: dns resolve s3-us-west-1.amazonaws.com:443 error: no record found for Query { name: Name("s3-us-west-1.amazonaws.com."), query_type: AAAA, query_class: IN }

ERROR tcp tunnel 127.0.0.1:18422 -> github-production-upload-manifest-file-7fdce7.s3.amazonaws.com:443 connect failed, error: dns resolve github-production-upload-manifest-file-7fdce7.s3.amazonaws.com:443 error: no record found for Query { name: Name("github-production-upload-manifest-file-7fdce7.s3.amazonaws.com."), query_type: AAAA, query_class: IN }

实际这两个域名是没有ipv6的，只有v4，所以会导致网页打不开

后经尝试，我把s3.amazonaws.com返回的ipv4限制到20个不会出错了，可以正常浏览访问

但是不知道这里出错是因为ip数量原因还是总体字节大小一类的

保险起见，我现在把返回ip数量限制在10个了，v4和v6最大各10各，目前一切正常

还请大佬给看看这怎么个事啊 

@zonyitoo

**Comment by zonyitoo** [comment]
The error log was reported originally from these calls:

https://github.com/shadowsocks/shadowsocks-rust/blob/ac459c753b0b4fca921fc80f7868f1861c5591c4/crates/shadowsocks/src/dns_resolver/resolver.rs#L375-L394

The first call is to use the system configured DNS server, servers in `/etc/resolv.conf` on Linux Platform. The second one is resolve with user's configured `dns` server.

These two paths are all using hickory resolver's `lookup_ip`, which was reporting:

1. `no record found for Query { name: Name("s3-us-west-1.amazonaws.com."), query_type: AAAA, query_class: IN }`
2. `no record found for Query { name: Name("github-production-upload-manifest-file-7fdce7.s3.amazonaws.com."), query_type: AAAA, query_class: IN }`

**Comment by zonyitoo** [comment]
You may try to rerun the program with `-vvv` to open all logs and see if we can find any error logs in hickory-dns.

**Comment by zonyitoo** [comment]
Interestingly I can't reproduce it locally, for example:

```
; <<>> DiG 9.11.26-RedHat-9.11.26-4.tl3 <<>> s3-us-west-1.amazonaws.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 37523
;; flags: qr rd ra; QUERY: 1, ANSWER: 8, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4000
;; QUESTION SECTION:
;s3-us-west-1.amazonaws.com.	IN	A

;; ANSWER SECTION:
s3-us-west-1.amazonaws.com. 5	IN	A	52.219.193.96
s3-us-west-1.amazonaws.com. 5	IN	A	52.219.116.104
s3-us-west-1.amazonaws.com. 5	IN	A	52.219.112.64
s3-us-west-1.amazonaws.com. 5	IN	A	52.219.193.152
s3-us-west-1.amazonaws.com. 5	IN	A	52.219.194.160
s3-us-west-1.amazonaws.com. 5	IN	A	52.219.193.176
s3-us-west-1.amazonaws.com. 5	IN	A	52.219.194.8
s3-us-west-1.amazonaws.com. 5	IN	A	52.219.193.8

;; Query time: 117 msec
;; SERVER: ...#53(...)
;; WHEN: Mon Feb 05 16:54:28 CST 2024
;; MSG SIZE  rcvd: 183
```

and

```
; <<>> DiG 9.11.26-RedHat-9.11.26-4.tl3 <<>> github-production-upload-manifest-file-7fdce7.s3.amazonaws.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 49512
;; flags: qr rd ra; QUERY: 1, ANSWER: 10, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4000
;; QUESTION SECTION:
;github-production-upload-manifest-file-7fdce7.s3.amazonaws.com.	IN A

;; ANSWER SECTION:
github-production-upload-manifest-file-7fdce7.s3.amazonaws.com.	42821 IN CNAME s3-1-w.amazonaws.com.
s3-1-w.amazonaws.com.	16	IN	CNAME	s3-w.us-east-1.amazonaws.com.
s3-w.us-east-1.amazonaws.com. 55 IN	A	52.217.173.137
s3-w.us-east-1.amazonaws.com. 55 IN	A	54.231.139.57
s3-w.us-east-1.amazonaws.com. 55 IN	A	52.217.14.212
s3-w.us-east-1.amazonaws.com. 55 IN	A	16.182.32.193
s3-w.us-east-1.amazonaws.com. 55 IN	A	3.5.17.230
s3-w.us-east-1.amazonaws.com. 55 IN	A	54.231.168.9
s3-w.us-east-1.amazonaws.com. 55 IN	A	52.217.92.20
s3-w.us-east-1.amazonaws.com. 55 IN	A	52.217.90.124

;; Query time: 27 msec
;; SERVER: ...#53(...)
;; WHEN: Mon Feb 05 16:55:02 CST 2024
;; MSG SIZE  rcvd: 269
```

It didn't return tons of IPs like your case.

**Comment by zonyitoo** [comment]
If smartdns could respond to hickory-dns's Query, we should first make a reproducible test case.

I couldn't test it locally, so expecting your test report.

**Comment by zonyitoo** [comment]
No worry, since your service working fine right now, there is no hurry.

Opening ports to strangers is not safe.

Checklists:

1. Just run `sslocal` or `ssserver` with `-vvv` and see if we can find any hickory-dns's error logs.
2. Test with hickory-dns's tool: https://github.com/hickory-dns/hickory-dns/blob/main/bin/src/hickory-dns.rs
3. Test with different protocols, for example, TCP. `smartdns` should support TCP DNS queries.

**Comment by zonyitoo** [comment]
Yes, try to set in configuration:

```json
{
    "dns": "tcp://127.0.0.1"
}
```

to force hickory-dns to use TCP queries.

> 当然，如果还不行就把ss rust服务端和客户端都加上vvv然后脱敏发过来

It should help.

**Comment by zonyitoo** [comment]
I just tried smartdns and set several DNS servers, but still couldn't get lots of IP reponses like your case.

**Comment by zonyitoo** [comment]
That's quite anonying that there was not even one line of hickory-dns's log found in the log, just simply returned `NoRecordsFound`.

> Feb 07 04:55:45 JP384 ssserver[17273]: time="2024-02-07T04:55:45+08:00" level=fatal msg="unable to initialise server state: unable to parse RedirAddr: unable to resolve RedirAddr: lookup www.那个网址.com on 127.0.0.1:53: read udp 127.0.0.1:40622->127.0.0.1:53: read: connection refused. "
> Feb 07 04:55:45 JP384 ssserver[17271]: 2024-02-07T04:55:45.009711212+08:00 ERROR plugin exited with status: exit status: 1

It shouldn't related to the `"dns"` configuration. Your plugin exited while server started.

Since it was "connection refused", did your smartdns enabled tcp listener?

Beyond that, there was no useful information could be found in these log. The bug is still remain unknown. :(

**Comment by zonyitoo** [comment]
https://github.com/shadowsocks/shadowsocks-rust/blob/7cec43908bdbe7beccb5dc51959293f0ca8b3518/crates/shadowsocks-service/src/dns/mod.rs#L24-L30

Please also try to set this environment variable: `SS_SYSTEM_DNS_RESOLVER_FORCE_BUILTIN` to use the libc provided DNS resolver.

**Comment by zonyitoo** [comment]
Yes, I have seen those comments.

> ```
> 2024-02-07T05:31:50.716172+08:00 JP384 ssserver[482]: 2024-02-07T05:31:50.715393849+08:00 TRACE [482:140665008683832] [shadowsocks_service::server::tcprelay] accepted tcp client connection 127.0.0.1:30340, establishing tunnel to s3.amazonaws.com:80
> 2024-02-07T05:32:05.720586+08:00 JP384 ssserver[482]: 2024-02-07T05:32:05.720165711+08:00 TRACE [482:140665008683832] [shadowsocks::dns_resolver::resolver] DNS resolved s3.amazonaws.com:80 with hickory-dns 15.00474s
> 2024-02-07T05:32:05.720672+08:00 JP384 ssserver[482]: 2024-02-07T05:32:05.720230684+08:00 ERROR [482:140665008683832] [shadowsocks_service::server::tcprelay] tcp tunnel 127.0.0.1:30340 -> s3.amazonaws.com:80 connect failed, error: dns resolve s3.amazonaws.com:80 error: no record found for Query { name: Name("s3.amazonaws.com."), query_type: AAAA, query_class: IN }
> 2024-02-07T05:32:05.720688+08:00 JP384 ssserver[482]: 2024-02-07T05:32:05.720296269+08:00 DEBUG [482:140665008683832] [shadowsocks_service::server::tcprelay] tcp server stream aborted with error: dns resolve s3.amazonaws.com:80 error: no record found for Query { name: Name("s3.amazonaws.com."), query_type: AAAA, query_class: IN }
> 
> 
> 
> 
> 2024-02-07T05:41:59.750190+08:00 JP384 ssserver[482]: 2024-02-07T05:41:59.748880727+08:00 TRACE [482:140665008683832] [shadowsocks_service::server::tcprelay] accepted tcp client connection 127.0.0.1:26532, establishing tunnel to www.youtube.com:443
> 2024-02-07T05:41:59.760460+08:00 JP384 ssserver[482]: 2024-02-07T05:41:59.760235021+08:00 TRACE [482:140665008683832] [shadowsocks::dns_resolver::resolver] DNS resolved www.youtube.com:443 with hickory-dns 0.011318416s
> 2024-02-07T05:41:59.760505+08:00 JP384 ssserver[482]: 2024-02-07T05:41:59.760273765+08:00 TRACE [482:140665008683832] [shadowsocks::net::tcp] trying connect www.youtube.com:443 172.217.175.46:443
> 2024-02-07T05:41:59.761803+08:00 JP384 ssserver[482]: 2024-02-07T05:41:59.761530892+08:00 TRACE [482:140665008683832] [shadowsocks::net::tcp] connected www.youtube.com:443 172.217.175.46:443
> 2024-02-07T05:41:59.761827+08:00 JP384 ssserver[482]: 2024-02-07T05:41:59.761608860+08:00 DEBUG [482:140665008683832] [shadowsocks_service::server::tcprelay] established tcp tunnel 127.0.0.1:26532 <-> www.youtube.com:443 with ConnectOpts { fwmark: None, bind_local_addr: None, bind_interface: None, tcp: TcpSocketOpts { send_buffer_size: None, recv_buffer_size: None, nodelay: false, fastopen: false, keepalive: Some(15s), mptcp: false }, udp: UdpSocketOpts { mtu: None } }
> ```
> 
> 再来一截子服务端的，还有个正常的youtube
> 
> 这个锅我琢磨可以甩给hickory-dns了

The key issue was in `ssserver`, there was no hickory-dns Query logs.

**Comment by zonyitoo** [comment]
For example:

```bash
export SS_SYSTEM_DNS_RESOLVER_FORCE_BUILTIN=true
ssserver -c config.json
```

**Comment by zonyitoo** [comment]
You have to remove the `"dns"` field in configuration file to let it use the “system” resolver.

**Comment by zonyitoo** [comment]
Then we can be sure that hickory-dns have problems when parsing those responses that have many IPs.

**Comment by zonyitoo** [comment]
Did you tried: https://github.com/hickory-dns/hickory-dns/blob/main/bin/src/hickory-dns.rs ? If it failed, we can open an issue in hickory-dns with this reproduciable case.

**Comment by zonyitoo** [comment]
Just walk through hickory-dns' source code, I found that they are using [tracing](https://crates.io/crates/tracing) framework to emit their debug logs.

I just added a simple tracing subscriber, please compile one and run with `-vvv`, and you will see hickory-dns' debug information.

```
2024-02-08T02:23:21.371826+08:00 DEBUG [hickory_proto::xfer] enqueueing message:QUERY:[Query { name: Name("s3.amazonaws.com"), query_type: AAAA, query_class: IN }] enqueueing message:QUERY:[Query { name: Name("s3.amazonaws.com"), query_type: AAAA, query_class: IN }]
2024-02-08T02:23:21.372119+08:00 DEBUG [hickory_proto::udp::udp_client_stream] final message: ; header 20519:QUERY:RD:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN A
 final message: ; header 20519:QUERY:RD:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN A

2024-02-08T02:23:21.37212+08:00 DEBUG [hickory_proto::udp::udp_client_stream] final message: ; header 15515:QUERY:RD:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN A
 final message: ; header 15515:QUERY:RD:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN A

2024-02-08T02:23:21.372204+08:00 DEBUG [hickory_proto::udp::udp_client_stream] final message: ; header 61911:QUERY:RD:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN AAAA
 final message: ; header 61911:QUERY:RD:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN AAAA

2024-02-08T02:23:21.372209+08:00 DEBUG [hickory_proto::udp::udp_client_stream] final message: ; header 22537:QUERY:RD:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN AAAA
 final message: ; header 22537:QUERY:RD:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN AAAA

2024-02-08T02:23:21.372328+08:00 DEBUG [hickory_proto::udp::udp_stream] created socket successfully created socket successfully
2024-02-08T02:23:21.372472+08:00 DEBUG [hickory_proto::udp::udp_stream] created socket successfully created socket successfully
2024-02-08T02:23:21.372568+08:00 DEBUG [hickory_proto::udp::udp_stream] created socket successfully created socket successfully
2024-02-08T02:23:21.372657+08:00 DEBUG [hickory_proto::udp::udp_stream] created socket successfully created socket successfully
2024-02-08T02:23:21.376149+08:00 DEBUG [hickory_proto::udp::udp_client_stream] received message id: 61911 received message id: 61911
2024-02-08T02:23:21.376237+08:00 DEBUG [hickory_resolver::error] Response:; header 61911:RESPONSE:RD,RA:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN AAAA
; answers 0
; nameservers 0
; additionals 0
 Response:; header 61911:RESPONSE:RD,RA:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN AAAA
; answers 0
; nameservers 0
; additionals 0

2024-02-08T02:23:21.376378+08:00 DEBUG [hickory_proto::udp::udp_client_stream] received message id: 20519 received message id: 20519
2024-02-08T02:23:21.37641+08:00 DEBUG [hickory_resolver::error] Response:; header 20519:RESPONSE:RD,RA:NoError:QUERY:8/0/0
; query
;; s3.amazonaws.com. IN A
; answers 8
s3.amazonaws.com. 547 IN A 52.216.147.85
s3.amazonaws.com. 547 IN A 52.217.118.64
s3.amazonaws.com. 547 IN A 52.216.56.64
s3.amazonaws.com. 547 IN A 52.217.162.152
s3.amazonaws.com. 547 IN A 52.216.217.56
s3.amazonaws.com. 547 IN A 54.231.225.152
s3.amazonaws.com. 547 IN A 52.217.114.40
s3.amazonaws.com. 547 IN A 52.217.175.48
; nameservers 0
; additionals 0
 Response:; header 20519:RESPONSE:RD,RA:NoError:QUERY:8/0/0
; query
;; s3.amazonaws.com. IN A
; answers 8
s3.amazonaws.com. 547 IN A 52.216.147.85
s3.amazonaws.com. 547 IN A 52.217.118.64
s3.amazonaws.com. 547 IN A 52.216.56.64
s3.amazonaws.com. 547 IN A 52.217.162.152
s3.amazonaws.com. 547 IN A 52.216.217.56
s3.amazonaws.com. 547 IN A 54.231.225.152
s3.amazonaws.com. 547 IN A 52.217.114.40
s3.amazonaws.com. 547 IN A 52.217.175.48
; nameservers 0
; additionals 0

2024-02-08T02:23:21.376524+08:00 DEBUG [hickory_resolver::error] Response:; header 20519:RESPONSE:RD,RA:NoError:QUERY:8/0/0
; query
;; s3.amazonaws.com. IN A
; answers 8
s3.amazonaws.com. 547 IN A 52.216.147.85
s3.amazonaws.com. 547 IN A 52.217.118.64
s3.amazonaws.com. 547 IN A 52.216.56.64
s3.amazonaws.com. 547 IN A 52.217.162.152
s3.amazonaws.com. 547 IN A 52.216.217.56
s3.amazonaws.com. 547 IN A 54.231.225.152
s3.amazonaws.com. 547 IN A 52.217.114.40
s3.amazonaws.com. 547 IN A 52.217.175.48
; nameservers 0
; additionals 0
 Response:; header 20519:RESPONSE:RD,RA:NoError:QUERY:8/0/0
; query
;; s3.amazonaws.com. IN A
; answers 8
s3.amazonaws.com. 547 IN A 52.216.147.85
s3.amazonaws.com. 547 IN A 52.217.118.64
s3.amazonaws.com. 547 IN A 52.216.56.64
s3.amazonaws.com. 547 IN A 52.217.162.152
s3.amazonaws.com. 547 IN A 52.216.217.56
s3.amazonaws.com. 547 IN A 54.231.225.152
s3.amazonaws.com. 547 IN A 52.217.114.40
s3.amazonaws.com. 547 IN A 52.217.175.48
; nameservers 0
; additionals 0

2024-02-08T02:23:21.376938+08:00 DEBUG [hickory_proto::udp::udp_client_stream] received message id: 22537 received message id: 22537
2024-02-08T02:23:21.376976+08:00 DEBUG [hickory_resolver::error] Response:; header 22537:RESPONSE:RD,RA:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN AAAA
; answers 0
; nameservers 0
; additionals 0
 Response:; header 22537:RESPONSE:RD,RA:NoError:QUERY:0/0/0
; query
;; s3.amazonaws.com. IN AAAA
; answers 0
; nameservers 0
; additionals 0

2024-02-08T02:23:21.377016+08:00 DEBUG [hickory_resolver::lookup_ip] one of ipv4 or ipv6 lookup failed in ipv4_and_ipv6 strategy: no record found for Query { name: Name("s3.amazonaws.com."), query_type: AAAA, query_class: IN } one of ipv4 or ipv6 lookup failed in ipv4_and_ipv6 strategy: no record found for Query { name: Name("s3.amazonaws.com."), query_type: AAAA, query_class: IN }
2024-02-08T02:23:21.377062+08:00 TRACE [64522:16971518] [shadowsocks::dns_resolver::resolver] DNS resolved s3.amazonaws.com:80 with hickory-dns 0.010445871s
```

**Comment by zonyitoo** [comment]
It looks like `edns` was default to be `false` (disabled) in hickory-dns resolver, which is why it rejected your large Response.

**Comment by zonyitoo** [comment]
@qwerttvv Please help to test if this commit fixes your issue.

**Comment by zonyitoo** [comment]
But even EDNS0 was disabled, TCP queries should be able to support large Response Messages. Emmm...

**Comment by zonyitoo** [comment]
According to your log, there is no bug.

```
one of ipv4 or ipv6 lookup failed in ipv4_and_ipv6 strategy: no record found for Query { name: Name("s3.amazonaws.com."), query_type: AAAA, query_class: IN }
```

This line indicated that `s3.amazonaws.com.` has no `AAAA` response.

```
trying connect s3.amazonaws.com:443 16.182.109.80:443
```

shadowsocks-rust got one of the resolved result `16.182.109.80`.

So this issue is now resolved.

**Comment by zonyitoo** [comment]
Nah. They are related. DNS specification uses 512 bytes as default UDP package size. 

https://en.wikipedia.org/wiki/Extension_Mechanisms_for_DNS

You can test with `dig s3.amazonaws.com` and see if you can find anything like this:

```
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 4096
```

**Comment by zonyitoo** [comment]
In hickory-dns' source code, it defines a constant 512 bytes as the UDP socket receive buffer's default size. The size of buffer can only be modified by the `Message`, which reminded me that there should be some place to configure this value, and then I found `edns0 = false` by default.

If software uses libc's provided DNS resolver, which in most case provided by the OS, should already enabled EDNS0.

Here are some possible cases that would have the same issue like we are discussing here:

1. Customized DNS resolver, like shadowsocks-libev, uses c-ares as its DNS resolver.
2. MUSL
...
